### ISSUE TITLE
Does 2.5.3 "Label in Name" apply to placeholder (where used as only visible label)?

### ISSUE THREAD

**Issue by detlevhfischer** [w3c/wcag#1700]
We are discussing whether 2.5.3 should also apply to fields which are visibly labelled only by placeholder text.
Scenario 1:

`<input type="text" aria-label="search term" placeholder="search">` 
A: 2.5.3 is not applicable since there is no (permanently) visible label (just placeholder) - N.A.
B:  2.5.3 applies since placeholder acts as visible label - PASS since visible label is contained in accName (provided via aria-label)

Scanario 2:
`<input type="text" placeholder="search">` 
A: 2.5.3 is not applicable since there is no (permanently) visible label (just placeholder) - N.A.
B:  2.5.3 applies since placeholder acts as visible label - FAIL since placeholder does not count as accName
B:  2.5.3 applies since placeholder acts as visible label - PASS since (most) browsers use placeholder as accName (in the absence of other sources)

**Comment by alastc** [comment]
I thought (but might need to lookup) that we decided placeholders didn't fulfil Labels and Instructions? (Assuming there are no instructions.) 

In which case: If placeholder doesn't count as a label, it would be logical for it to be NA for both scenarios.

**Comment by awkawk** [comment]
I agree @alastc - placeholder text isn't to be used for meeting label in name.

@detlevhfischer no WCAG SC are Not Applicable, you either pass or you don't (I'm sure that you are just using Not Applicable as a shorthand, but just in case...). If the situation doesn't come up the you pass by not failing. For example, you pass 1.2.3 for audio description when you don't have video.

For 2.5.3, it starts with "For user interface components with labels that include text or images of text" so if you have a control that is labeled by an image, 2.5.3 is passed because there isn't a label with text or images of text.

The odd situation I think of is where there is an apparent placeholder but when you focus the input the placeholder text appears above the input so it is still visible. I've seen this done such as in Material Design Lite (https://getmdl.io/components/index.html#textfields-section) but the placeholder is actually the label element properly associated with the control. If someone implemented this so that the actual placeholder attribute was used and then the label appeared when the input was focused then I supposed that the placeholder would be the label for the unfocused state and would need to meet 2.5.3, but that seems like a more convoluted implementation.

**Comment by JAWS-test** [comment]
> FAIL since placeholder does not count as accName

That's not true, see: https://www.w3.org/TR/html-aam-1.0/#input-type-text-input-type-password-input-type-search-input-type-tel-input-type-url-and-textarea-element:

> Otherwise use the control's placeholder attribute.

**Comment by JAWS-test** [comment]
> The odd situation I think of is where there is an apparent placeholder but when you focus the input the placeholder text appears above the input so it is still visible.

There are many frameworks that do this. I don't find that odd. It is then simply that the label (not the placeholder) is visually moved into the input field and set over the field when receiving focus or after text input. In this case I would consider 2.5.3 applicable because there is no placeholder but a label

**Comment by jake-abma** [comment]
@awk wrote:

> I agree @alastc - placeholder text isn't to be used for meeting label in name.

The SC Understanding says:

> Note that placeholder text within an input field is not considered an appropriate means of providing a label. The HTML5 specification states The placeholder attribute should not be used as an alternative to a <label>. However, it is worth noting that "label" in that HTML5 statement is in code brackets and links to the label element. For the purposes of this Label in Name Success Criterion, "label" is not used in such a programmatic sense but is simply referring to a text string in close visual proximity to a component. As such, in the absence of any other nearby text string (as described in the preceding list), if an input contains placeholder text, such text may be a candidate for Label in Name. This is supported both through the accessible name calculation (discussed later) and from the practical sense that where a visible label is not otherwise provided, it is likely that a speech-input user may attempt to use the placeholder text value as a means of interacting with the input.

Please note: 

> **As such, in the absence of any other nearby text string (as described in the preceding list), if an input contains placeholder text, such text may be a candidate for Label in Name.**

**Comment by jake-abma** [comment]
Also the definition of label doesn't exclude placeholder text...

https://www.w3.org/WAI/WCAG22/Understanding/label-in-name.html#dfn-label

> text or other component with a text alternative that is presented to a user to identify a component within Web content

**Comment by jlukosky** [comment]
It sounds like placeholder as an accessible name is like painting your house with a rag instead of a brush. It's "inappropriate" (agreed) but.... it works.

**Comment by alastc** [comment]
> As such, in the absence of any other nearby text string (as described in the preceding list), if an input contains placeholder text, such text may be a candidate for Label in Name.

Ok, so we have a slight conflict for a particular scenario, where in general we consider disappearing things to not be sufficient, e.g. for label & instructions.

To put up an alternative, how about adjusting the [quoted paragraph above](https://github.com/w3c/wcag/issues/1700#issuecomment-808893420) to:

> For the purposes of this Label in Name Success Criterion, "label" is not used in such a programmatic sense but is simply referring to a text string in close visual proximity to a component. As such, if the visual label disappears when you interact with the input, it is not a permanent label for the element.
> 
> The placeholder may provide the accessible name, and it will be seen by the user first (assuming there is not data present in the input), but the temporary nature of the placeholder means that it is not considered a 'label'.

**Comment by davvwar** [comment]
This is similar to points I've come across with 2.4.6: Headings and Labels and 3.3.2: Labels or Instructions. That is, the "Label" in this context is the permanent visual signifier that communicates the purpose of an adjacent form field. 

It may or may not be done as a `<label>`. It may or may not be part of the accessible name. But those are the remit of other success criterion, such as 1.3.1: Info and Relationships and 4.1.2: Name, Role, Value. 

A search could pass 2.4.6: Headings and Labels by just having a magnifying glass icon permanently pinned to it. It may fail other success criteria, based on the accessible name of the field, but that is a separate matter. 

In the context of 2.5.3: Label in Name, my understanding is that:

- if the field has a label (i.e. a permanent visual signifier), AND
- if the permanent visual signifier has text, THEN
- the text of the label (i.e. the permanent visual signifier) must be included in the accessible name. 

In this context, the placeholder may or may not be mapped to the accessible name. However, it is not a "label" when we apply the definition of a permanent visual signifier.

It is arguably first and foremost a 3.3.2: Labels or Instructions issue when placeholder is used as the only visible label. It arguably does not fail 2.5.3: Label in Name in as much as there is no "label" to be included in the accessible name.
